### PR TITLE
cluster,dm:  auto detect cpu arch

### PR DESCRIPTION
**Pull request by 9547**
### What problem does this PR solve? 

Implement #1397 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
NONE
```

### CODE REVIEW

File: pkg/cluster/manager/manager.go [hunk 1]
```diff
@@ -134,3 +136,54 @@ func (m *Manager) sshTaskBuilder(name string, topo spec.Topology, user string, o
 		).
 		ClusterSSH(topo, user, opts.SSHTimeout, opts.SSHType, topo.BaseTopo().GlobalOptions.SSHType)
 }
+
+func (m *Manager) fillHostArch(s *cliutil.SSHConnectionProps, topo spec.Topology, gOpt *operator.Options, user string) error {
+	hostArch := map[string]string{}
```
Comment (AstroProfundis) [inline pkg/cluster/manager/manager.go:141]: SSH port should also be part of the unique key, the SSH target might be a NAT device.

File: pkg/cluster/manager/manager.go [hunk 2]
```diff
@@ -134,3 +136,54 @@ func (m *Manager) sshTaskBuilder(name string, topo spec.Topology, user string, o
...
+	t := task.NewBuilder().
+		ParallelStep("+ Detect CPU Arch", false, detectTasks...).
+		Build()
+
+	if err := t.Execute(ctx); err != nil {
+		return perrs.Annotate(err, "failed to fetch cpu arch")
+	}
+
+	for host := range hostArch {
+		stdout, _, ok := ctxt.GetInner(ctx).GetOutputs(host)
```
Comment (AstroProfundis) [inline pkg/cluster/manager/manager.go:182]: ... and so the result handling should be just after the probe, as there might be multiple job for one host, it's possible to append a `Func` task after `Shell` to do so.

File: pkg/cluster/spec/spec.go [hunk 3]
```diff
@@ -71,7 +71,7 @@ type (
 		LogDir          string               `yaml:"log_dir,omitempty"`
 		ResourceControl meta.ResourceControl `yaml:"resource_control,omitempty" validate:"resource_control:editable"`
 		OS              string               `yaml:"os,omitempty" default:"linux"`
-		Arch            string               `yaml:"arch,omitempty" default:"amd64"`
+		Arch            string               `yaml:"arch,omitempty" default:""`
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec.go:74]: just remove the `default` tag and it will be zero value of the type.

**Review by AstroProfundis — COMMENTED** [review]
What if user specified an arch different with `uname -m` reports? e.g., user sets `amd64` in the topology file, but the probe shows as `aarch64`. Will the topology changed by the probe result?

**Comment by 9547** [comment]
> What if user specified an arch different with `uname -m` reports? e.g., user sets `amd64` in the topology file, but the probe shows as `aarch64`. Will the topology changed by the probe result?

We only detect those nodes whose arch was not specified.

File: pkg/cluster/manager/manager.go (hunk 1 above)
Comment (9547) [reply pkg/cluster/manager/manager.go:141]: When reviewing the rest of the code, I found that for the most part, we didn't handle the SSH Port case. 

eg. Shell's execute will set output by the host, if two nodes are under the same host, the later one's output will be overwritten. 

https://github.com/pingcap/tiup/blob/342f0448f9225ce8b87238319e93a659e1e71188/pkg/cluster/task/shell.go#L34-L53

File: pkg/cluster/manager/manager.go (hunk 2 above)
Comment (9547) [reply pkg/cluster/manager/manager.go:182]: What do you mean is make this task along with the later ones, such as check, deploy, scale-out? If so, IMO, it is best to have as little impact on the existing business as possible, and the CPU detection is a pre-action that needs to be done before the main logic(check, deploy, scale-out) is executed.  So, here I've implemented this action in an independent operation.

**Review by AstroProfundis — APPROVED** [review]

File: pkg/cluster/manager/manager.go (hunk 2 above)
Comment (AstroProfundis) [reply pkg/cluster/manager/manager.go:181]: Nope, I mean appending a `Func` task just after the `Shell` task that runs `uname`, but as we don't distinguish host by SSH port, it not necessary.

**Pull request merged**